File: backend/search/score.go
// Package search 是全局搜索的内核：模糊打分（score.go）与项目文件名索引（index.go）。
//
// 打分照 fzf 的思路做了一版小的：**子序列匹配 + 位置奖励 + 间隔惩罚**。纯子串包含
// 不够用——`ovw` 要能命中 `Overview.tsx`、`src/ap.go` 要能命中 `backend/api/api.go`；
// 而纯子序列又会把 `a...b...c` 这种跨半个路径的巧合排到真正的词首命中前面，所以
// 命中在词首（`/ _ - . 空格`之后）、驼峰拐点、连续命中都要加分，中间跳过的字符要
// 扣分。返回命中位置是为了让 UI 把匹配的字符加粗——搜索结果不标出「为什么它匹配」，
// 用户就只能靠猜。
package search

import (
	"sort"
	"strings"
	"unicode"
)

const (
	scoreMatch       = 16 // 每命中一个字符的底分
	bonusBoundary    = 14 // 词首：分隔符之后（路径段开头、下划线/连字符后）
	bonusCamel       = 10 // 驼峰拐点 aB、字母后接数字
	bonusConsecutive = 12 // 与上一个命中字符相邻
	bonusFirstChar   = 12 // 整串的第一个字符就命中
	penaltyGapStart  = 3  // 跳过一段的起步惩罚
	penaltyGapExtend = 1  // 跳过的每个额外字符
	penaltyLateStart = 12 // 首个命中离串首越远越弱，封顶这个数

	// maxTargetRunes 单个目标参与打分的长度上限。路径再长也只看前 256 个字符：
	// DP 是 O(词长 × 目标长)，不封顶的话一条超长路径就能把一次搜索拖慢一个量级。
	maxTargetRunes = 256

	negInf = -1 << 30
)

// Match 是一次匹配的结果。Positions 是命中字符在**原始目标串**上的 rune 下标（升序），
// 给 UI 高亮用；多词查询时是各词命中位置的并集。
type Match struct {
	Score     int   `json:"score"`
	Positions []int `json:"positions,omitempty"`
}

// Query 是编译好的查询：按空白切成若干词，每个词预先转成小写 rune。
// 一次搜索要拿同一个查询去打几万个候选，切词和转 rune 只该做一次。
type Query struct {
	terms [][]rune
}

// Compile 把查询串编译成 Query。多个词是 AND 关系——「api search」要求两个词都命中。
func Compile(query string) Query {
	fields := strings.Fields(strings.ToLower(query))
	terms := make([][]rune, 0, len(fields))
	for _, f := range fields {
		terms = append(terms, []rune(f))
	}
	return Query{terms: terms}
}

// Empty 报告查询是否为空（空查询一律不匹配，由调用方决定要不要给「全部」）。
func (q Query) Empty() bool { return len(q.terms) == 0 }

// CouldMatch 是廉价预筛：只做「每个词都是目标的子序列」的判断，不进 DP、不分配。
// 打几万个候选时先过这一关，真正进 DP 的通常只剩百分之几。
func (q Query) CouldMatch(target string) bool {
	if len(q.terms) == 0 {
		return false
	}
	for _, term := range q.terms {
		i := 0
		for _, r := range target {
			if unicode.ToLower(r) == term[i] {
				i++
				if i == len(term) {
					break
				}
			}
		}
		if i < len(term) {
			return false
		}
	}
	return true
}

// Score 给 query 对 target 的匹配打分。query 按空白切词做 AND：每个词都要命中，
// 总分相加——这样「roam 概览」「api search」这种两段式查询才有意义。
// 不匹配返回 ok=false（0 分和「勉强匹配」要能区分开）。
func Score(query, target string) (Match, bool) { return Compile(query).Score(target) }

// Score 用已编译的查询打分，语义同包级 Score。
func (q Query) Score(target string) (Match, bool) {
	if len(q.terms) == 0 || target == "" {
		return Match{}, false
	}
	// 先做零分配的预筛。一次搜索要打几十万个候选，其中 99% 连子序列都不是；
	// 不先挡一道的话，光是「转 rune + 算位置奖励」两次分配就够把延迟拖到半秒。
	if !q.CouldMatch(target) {
		return Match{}, false
	}
	runes := []rune(target)
	if len(runes) > maxTargetRunes {
		runes = runes[:maxTargetRunes]
	}
	lower := make([]rune, len(runes))
	for i, r := range runes {
		lower[i] = unicode.ToLower(r)
	}
	bonus := boundaryBonuses(runes)

	total := 0
	hit := map[int]bool{}
	for _, term := range q.terms {
		s, pos, ok := scoreTerm(term, lower, bonus)
		if !ok {
			return Match{}, false
		}
		total += s
		for _, p := range pos {
			hit[p] = true
		}
	}
	positions := make([]int, 0, len(hit))
	for p := range hit {
		positions = append(positions, p)
	}
	sort.Ints(positions)
	return Match{Score: total, Positions: positions}, true
}

// Best 在多个字段里取最高分，返回命中的字段下标。副字段（下标 > 0）打折：
// 名字命中永远比路径命中更相关，不打折的话搜 `api` 会让一堆路径里带 api 的文件
// 压过真叫 api 的那个。
func Best(query string, fields ...string) (Match, int, bool) {
	return Compile(query).Best(fields...)
}

// Best 用已编译的查询在多个字段里取最高分，语义同包级 Best。
func (q Query) Best(fields ...string) (Match, int, bool) {
	best, bestIdx, found := Match{}, -1, false
	for i, f := range fields {
		m, ok := q.Score(f)
		if !ok {
			continue
		}
		if i > 0 {
			m.Score = m.Score * 3 / 4
		}
		if !found || m.Score > best.Score {
			best, bestIdx, found = m, i, true
		}
	}
	return best, bestIdx, found
}

// boundaryBonuses 预先算好每个位置的「位置奖励」：整串开头、分隔符之后、驼峰拐点。
func boundaryBonuses(t []rune) []int {
	b := make([]int, len(t))
	for i, r := range t {
		if i == 0 {
			b[i] = bonusBoundary
			continue
		}
		p := t[i-1]
		switch {
		case isSep(p):
			b[i] = bonusBoundary
		case unicode.IsLower(p) && unicode.IsUpper(r):
			b[i] = bonusCamel
		case unicode.IsDigit(r) && !unicode.IsDigit(p):
			b[i] = bonusCamel
		}
	}
	return b
}

func isSep(r rune) bool {
	switch r {
	case '/', '\\', '_', '-', '.', ' ', ':', '@', '(', '[':
		return true
	}
	return false
}

// scoreTerm 是单个词的 DP：cur[j] = 「词的第 i 个字符落在目标第 j 位」时的最高分。
// 前驱只有两种来源——紧邻的上一位（连续，加奖励）或更早的位置（跳过，按间隔扣分），
// 后者用一个随 j 衰减的滑动最大值维护，整体 O(词长 × 目标长)。
func scoreTerm(q, tLower []rune, bonus []int) (int, []int, bool) {
	n, m := len(tLower), len(q)
	if m == 0 || m > n {
		return 0, nil, false
	}
	if !isSubsequence(q, tLower) {
		return 0, nil, false // 连子序列都不是，不必进 DP
	}

	back := make([][]int32, m) // back[i][j] = 第 i-1 个字符落在哪一位，回溯高亮位置用
	prev := make([]int, n)
	cur := make([]int, n)

	for i := 0; i < m; i++ {
		back[i] = make([]int32, n)
		gap, gapIdx := negInf, -1 // 「跳过一段后接上」的最优前驱
		for j := 0; j < n; j++ {
			// 位置 j 的间隔候选只能来自 k <= j-2（k = j-1 属于连续那一支）
			if i > 0 && j >= 2 {
				if gap != negInf {
					gap -= penaltyGapExtend
				}
				if prev[j-2] > negInf/2 && prev[j-2]-penaltyGapStart > gap {
					gap, gapIdx = prev[j-2]-penaltyGapStart, j-2
				}
			}
			cur[j] = negInf
			back[i][j] = -1
			if q[i] != tLower[j] {
				continue
			}
			if i == 0 {
				s := scoreMatch + bonus[j]
				if j == 0 {
					s += bonusFirstChar
				}
				if j < penaltyLateStart {
					s -= j
				} else {
					s -= penaltyLateStart
				}
				cur[j] = s
				continue
			}
			best, bestIdx := negInf, -1
			if j > 0 && prev[j-1] > negInf/2 {
				best, bestIdx = prev[j-1]+bonusConsecutive, j-1
			}
			if gapIdx >= 0 && gap > best {
				best, bestIdx = gap, gapIdx
			}
			if bestIdx < 0 {
				continue
			}
			cur[j] = best + scoreMatch + bonus[j]
			back[i][j] = int32(bestIdx)
		}
		prev, cur = cur, prev
	}

	bestScore, bestEnd := negInf, -1
	for j := 0; j < n; j++ {
		if prev[j] > bestScore {
			bestScore, bestEnd = prev[j], j
		}
	}
	if bestEnd < 0 || bestScore <= negInf/2 {
		return 0, nil, false
	}
	pos := make([]int, m)
	j := bestEnd
	for i := m - 1; i >= 0; i-- {
		pos[i] = j
		j = int(back[i][j])
	}
	return bestScore, pos, true
}

func isSubsequence(q, t []rune) bool {
	i := 0
	for _, r := range t {
		if r == q[i] {
			i++
			if i == len(q) {
				return true
			}
		}
	}
	return false
}

File: frontend/src/components/search/score.ts
// 模糊打分：**子序列匹配 + 位置奖励 + 间隔惩罚**（backend/search/score.go 的等价实现）。
//
// 为什么前端也要有一份：结果分两路来（本地内存里的页面/已开会话，后端来的项目/
// 会话/文件），两路必须用同一套规则打分，否则合并后的排序自相矛盾——同样叫
// `overview` 的两条，一条排第一、一条排第七。
//
// 改这里就要同步改 Go 那份，两边的用例钉在 score.test.ts 与 score_test.go。
const SCORE_MATCH = 16
const BONUS_BOUNDARY = 14 // 词首：分隔符之后
const BONUS_CAMEL = 10 // 驼峰拐点、字母后接数字
const BONUS_CONSECUTIVE = 12 // 与上一个命中字符相邻
const BONUS_FIRST_CHAR = 12 // 整串第一个字符就命中
const PENALTY_GAP_START = 3 // 跳过一段的起步惩罚
const PENALTY_GAP_EXTEND = 1 // 跳过的每个额外字符
const PENALTY_LATE_START = 12 // 首个命中离串首越远越弱，封顶
const MAX_TARGET = 256 // 单个目标参与打分的长度上限
const NEG = -1e9

export type ScoreResult = { score: number; positions: number[] }

const SEP = new Set(['/', '\\', '_', '-', '.', ' ', ':', '@', '(', '['])
const isUpper = (c: string) => c !== c.toLowerCase() && c === c.toUpperCase()
const isLower = (c: string) => c !== c.toUpperCase() && c === c.toLowerCase()
const isDigit = (c: string) => c >= '0' && c <= '9'

function boundaryBonuses(t: string[]): number[] {
  return t.map((c, i) => {
    if (i === 0) return BONUS_BOUNDARY
    const p = t[i - 1]
    if (SEP.has(p)) return BONUS_BOUNDARY
    if (isLower(p) && isUpper(c)) return BONUS_CAMEL
    if (isDigit(c) && !isDigit(p)) return BONUS_CAMEL
    return 0
  })
}

/**
 * 单个词的 DP：cur[j] = 「词的第 i 个字符落在目标第 j 位」时的最高分。
 * 前驱只有两种来源——紧邻的上一位（连续，加奖励）或更早的位置（跳过，按间隔扣分），
 * 后者用一个随 j 衰减的滑动最大值维护，整体 O(词长 × 目标长)。
 */
function scoreTerm(q: string[], tLower: string[], bonus: number[]): ScoreResult | null {
  const n = tLower.length, m = q.length
  if (m === 0 || m > n) return null
  const back: Int32Array[] = []
  let prev = new Float64Array(n).fill(NEG)
  let cur = new Float64Array(n).fill(NEG)
  for (let i = 0; i < m; i++) {
    const row = new Int32Array(n).fill(-1)
    back.push(row)
    let gap = NEG, gapIdx = -1
    for (let j = 0; j < n; j++) {
      // 位置 j 的「间隔候选」只能来自 k <= j-2（k = j-1 属于连续那一支）
      if (i > 0 && j >= 2) {
        if (gap > NEG) gap -= PENALTY_GAP_EXTEND
        if (prev[j - 2] > NEG / 2 && prev[j - 2] - PENALTY_GAP_START > gap) {
          gap = prev[j - 2] - PENALTY_GAP_START
          gapIdx = j - 2
        }
      }
      cur[j] = NEG
      if (q[i] !== tLower[j]) continue
      if (i === 0) {
        let s = SCORE_MATCH + bonus[j]
        if (j === 0) s += BONUS_FIRST_CHAR
        s -= Math.min(j, PENALTY_LATE_START)
        cur[j] = s
        continue
      }
      let best = NEG, bestIdx = -1
      if (j > 0 && prev[j - 1] > NEG / 2) { best = prev[j - 1] + BONUS_CONSECUTIVE; bestIdx = j - 1 }
      if (gapIdx >= 0 && gap > best) { best = gap; bestIdx = gapIdx }
      if (bestIdx < 0) continue
      cur[j] = best + SCORE_MATCH + bonus[j]
      row[j] = bestIdx
    }
    const swap = prev; prev = cur; cur = swap
  }
  let bestScore = NEG, bestEnd = -1
  for (let j = 0; j < n; j++) if (prev[j] > bestScore) { bestScore = prev[j]; bestEnd = j }
  if (bestEnd < 0 || bestScore <= NEG / 2) return null
  const positions = new Array<number>(m)
  let j = bestEnd
  for (let i = m - 1; i >= 0; i--) { positions[i] = j; j = back[i][j] }
  return { score: bestScore, positions }
}

/** 模糊打分：查询按空白切词做 AND，返回总分与命中位置（不匹配返回 null） */
export function fuzzyScore(query: string, target: string): ScoreResult | null {
  const terms = query.toLowerCase().split(/\s+/).filter(Boolean)
  if (!terms.length || !target) return null
  const t = Array.from(target).slice(0, MAX_TARGET)
  const lower = t.map((c) => c.toLowerCase())
  const bonus = boundaryBonuses(t)
  let total = 0
  const hit = new Set<number>()
  for (const term of terms) {
    const r = scoreTerm(Array.from(term), lower, bonus)
    if (!r) return null
    total += r.score
    r.positions.forEach((p) => hit.add(p))
  }
  return { score: total, positions: [...hit].sort((a, b) => a - b) }
}

/**
 * 多字段取最高分；副字段（下标 > 0）打 75 折——名字命中永远比路径命中更相关，
 * 不打折的话搜 `api` 会让一堆路径里带 api 的文件压过真叫 api 的那个。
 * 位置只在主字段命中时给：在标题上标出根本没匹配的字符比不标更糊涂。
 */
export function fuzzyBest(query: string, fields: string[]): { score: number; positions?: number[] } | null {
  let best: { score: number; positions?: number[] } | null = null
  fields.forEach((f, i) => {
    const r = fuzzyScore(query, f)
    if (!r) return
    const score = i === 0 ? r.score : Math.floor((r.score * 3) / 4)
    if (!best || score > best.score) best = { score, positions: i === 0 ? r.positions : undefined }
  })
  return best
}
